### PR TITLE
Godot 4.0 Package

### PR DESCRIPTION
**Pull request by abueide**
###### Motivation for this change
I wanted to be able to build & run godot 4 from the master branch. This package with some fixes applied lets us track specific master commits & also prepare for the godot 4 alpha.

###### Things done
Copied new godot package to a new godot4 package.
Changed revision to latest master commit
Updated patch file to fix build for godot 4.0

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/godot4/default.nix [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+let
+  options = {
+    touch = libXi != null;
+    pulseaudio = false;
+    udev = withUdev;
+  };
+in stdenv.mkDerivation rec {
+  pname = "godot4";
+  version = "4.0-dev";
```
Comment (Mindavi) [inline pkgs/development/tools/godot4/default.nix:16]: I think the convention is 'unstable-<date>' for the version.

File: pkgs/development/tools/godot4/default.nix [hunk 2]
```diff
@@ -0,0 +1,71 @@
+{ stdenv, lib, fetchFromGitHub, scons, pkg-config, udev, libX11
+, libXcursor , libXinerama, libXrandr, libXrender, libpulseaudio
+, libXi, libXext, libXfixes, freetype, openssl
```
Comment (Mindavi) [inline pkgs/development/tools/godot4/default.nix:3]: Consider running nixpkgs-fmt on this.

File: pkgs/development/tools/godot4/pkg_config_additions.patch [hunk 3]
```diff
@@ -0,0 +1,30 @@
+diff --git a/platform/linuxbsd/detect.py b/platform/linuxbsd/detect.py
+index 1487210..394cb9e 100755
+--- a/platform/linuxbsd/detect.py
++++ b/platform/linuxbsd/detect.py
+@@ -214,6 +214,9 @@ def configure(env):
+     env.ParseConfig("pkg-config xrandr --cflags --libs")
+     env.ParseConfig("pkg-config xrender --cflags --libs")
+     env.ParseConfig("pkg-config xi --cflags --libs")
++    env.ParseConfig("pkg-config xfixes --cflags --libs")
++    env.ParseConfig("pkg-config glu --cflags --libs")
```
Comment (Mindavi) [inline pkgs/development/tools/godot4/pkg_config_additions.patch:10]: Is this also intended for upstreaming? Looks like it could be.

**Review by Mindavi — COMMENTED** [review]
Haven't tried this, but left some comments.

File: pkgs/development/tools/godot4/pkg_config_additions.patch (hunk 3 above)
Comment (abueide) [reply pkgs/development/tools/godot4/pkg_config_additions.patch:10]: @Mindavi i just updated the patch file in the current godot package. I'm not exactly sure why the patch is needed or if it should be upstreamed.

File: pkgs/development/tools/godot4/server.nix [hunk 4]
```diff
@@ -0,0 +1,18 @@
+{ godot, lib }:
+godot.overrideAttrs (oldAttrs: rec {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/server.nix:2]:
```suggestion
{ godot, lib }:

godot.overrideAttrs (oldAttrs: rec {
```

File: pkgs/development/tools/godot4/server.nix [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+  pname = "godot-server";
+  sconsFlags = "target=release platform=server tools=no";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/server.nix:4]: You are sure this is not an array?

File: pkgs/development/tools/godot4/server.nix [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+  installPhase = ''
+    mkdir -p "$out/bin"
+    cp bin/godot_server.* $out/bin/godot-server
+
+    mkdir "$dev"
+    cp -r modules/gdnative/include $dev
+
+    mkdir -p "$man/share/man/man6"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/server.nix:12]:
```suggestion
    mkdir -p $out/bin $dev
    cp bin/godot_server.* $out/bin/godot-server
    cp -r modules/gdnative/include $dev
```

File: pkgs/development/tools/godot4/server.nix [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+    cp misc/dist/linux/godot.6 "$man/share/man/man6/"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/server.nix:13]: This should use installManPage.

File: pkgs/development/tools/godot4/server.nix [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+  '';
+  meta.description =
+    "Free and Open Source 2D and 3D game engine (server build)";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/server.nix:16]:
```suggestion
  meta.description = "Free and Open Source 2D and 3D game engine (server build)";
```

File: pkgs/development/tools/godot4/headless.nix [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    cp bin/godot_server.* $out/bin/godot-headless
+
+    mkdir "$dev"
+    cp -r modules/gdnative/include $dev
+
+    mkdir -p "$man/share/man/man6"
+    cp misc/dist/linux/godot.6 "$man/share/man/man6/"
+  '';
+  meta.description =
+    "Free and Open Source 2D and 3D game engine (headless build)";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/headless.nix:16]:
```suggestion
  meta.description = "Free and Open Source 2D and 3D game engine (headless build)";
```

File: pkgs/development/tools/godot4/headless.nix [hunk 10]
```diff
@@ -0,0 +1,18 @@
...
+  sconsFlags = "target=release_debug platform=server tools=yes";
+  installPhase = ''
+    mkdir -p "$out/bin"
+    cp bin/godot_server.* $out/bin/godot-headless
+
+    mkdir "$dev"
+    cp -r modules/gdnative/include $dev
+
+    mkdir -p "$man/share/man/man6"
+    cp misc/dist/linux/godot.6 "$man/share/man/man6/"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/headless.nix:13]: Use installManPage.

File: pkgs/development/tools/godot4/headless.nix [hunk 11]
```diff
@@ -0,0 +1,18 @@
+{ godot, lib }:
+godot.overrideAttrs (oldAttrs: rec {
+  pname = "godot-headless";
+  sconsFlags = "target=release_debug platform=server tools=yes";
+  installPhase = ''
+    mkdir -p "$out/bin"
+    cp bin/godot_server.* $out/bin/godot-headless
+
+    mkdir "$dev"
+    cp -r modules/gdnative/include $dev
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/headless.nix:10]:
```suggestion
    mkdir -p $out/bin $dev
    cp bin/godot_server.* $out/bin/godot-headless
    cp -r modules/gdnative/include $dev
```

File: pkgs/development/tools/godot4/headless.nix [hunk 12]
```diff
@@ -0,0 +1,18 @@
+{ godot, lib }:
+godot.overrideAttrs (oldAttrs: rec {
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/headless.nix:2]:
```suggestion
{ godot, lib }:

godot.overrideAttrs (oldAttrs: rec {
```

File: pkgs/development/tools/godot4/export-templates.nix [hunk 13]
```diff
@@ -0,0 +1,17 @@
...
+  sconsFlags = "target=release platform=x11 tools=no";
+  installPhase = ''
+    # The godot export command expects the export templates at
+    # .../share/godot/templates/3.2.3.stable with 3.2.3 being the godot version.
+    mkdir -p "$out/share/godot/templates/${oldAttrs.version}.stable"
+    cp bin/godot.x11.opt.64 $out/share/godot/templates/${oldAttrs.version}.stable/linux_x11_64_release
+  '';
+  outputs = [ "out" ];
+  meta.description =
+    "Free and Open Source 2D and 3D game engine (export templates)";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/export-templates.nix:15]:
```suggestion
  meta.description = "Free and Open Source 2D and 3D game engine (export templates)";
```

File: pkgs/development/tools/godot4/default.nix [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+  '';
+
+  outputs = [ "out" "dev" "man" ];
+
+  installPhase = ''
+    mkdir -p "$out/bin"
+    cp bin/godot.* $out/bin/godot
+
+    mkdir "$dev"
+    cp -r modules/gdnative/include $dev
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/default.nix:51]:
```suggestion
    mkdir -p $out/bin $dev
    cp bin/godot.* $out/bin/godot
    cp -r modules/gdnative/include $dev
```

File: pkgs/development/tools/godot4/default.nix [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+
+    mkdir -p "$man/share/man/man6"
+    cp misc/dist/linux/godot.6 "$man/share/man/man6/"
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/default.nix:54]: Use installManPage.

File: pkgs/development/tools/godot4/default.nix [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    mkdir -p "$out"/share/{applications,icons/hicolor/scalable/apps}
+    cp misc/dist/linux/org.godotengine.Godot.desktop "$out/share/applications/"
+    cp icon.svg "$out/share/icons/hicolor/scalable/apps/godot.svg"
+    cp icon.png "$out/share/icons/godot.png"
+    substituteInPlace "$out/share/applications/org.godotengine.Godot.desktop" \
```
Comment (SuperSandro2000) [inline pkgs/development/tools/godot4/default.nix:60]:
```suggestion
    mkdir -p $out/share/{applications,icons/hicolor/scalable/apps}
    cp misc/dist/linux/org.godotengine.Godot.desktop $out/share/applications/
    cp icon.svg $out/share/icons/hicolor/scalable/apps/godot.svg
    cp icon.png $out/share/icons/godot.png
    substituteInPlace $out/share/applications/org.godotengine.Godot.desktop \
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/development/tools/godot4/server.nix (hunk 5 above)
Comment (abueide) [reply pkgs/development/tools/godot4/server.nix:4]: @SuperSandro2000 this file remains unchanged from the original godot package.

**Comment by Krutonium** [comment]
If this ever does get merged, I'd additionally love a version that does godot-mono

**Comment by wegank** [comment]
Closed by #196881.

**Pull request closed without merging**